### DISCUSSION TITLE
How to add externally captured requests and responses?

### DISCUSSION THREAD

**Discussion by lalitpagaria** [keploy/keploy#220] [Q&A]
First thank you for such a nice tool.

Mainly I liked the idea about identification of noisy fields and generating mock test accordingly.
I would like to use this capability to generate mock tests for our open-source tool: [Obsei](https://github.com/obsei/obsei).

We have lot of external integrations and like to mock them for testing purpose. Hence like to generate mocks programmatically as follows in github actions (till Keploy have python SDK)
- Collect 5-10 requests and responses of external integrations (Twitter, Slack etc) - Or ideally via CSV/TSV
- Push it to Keploy via API call (to find noisy fields)
- Download mocks
- Use these mocks in Obsei tests

Pardon my ignorance about project and for this naïve query :)

**Comment by slayerjain** [comment]
Thanks for the interest @lalitpagaria! 

To capture the mocks we can :
1. Write a simple go (the most supported SDK at the moment) application which calls the integrations
2. Integrate keploy SDK and record mocks
3. Use the generated YAML in any tests.

I think what's missing is that to use the mocks, you would need a python SDK (assuming you're testing a python app). The mocks are generated as yamls, and there has to be some python code which would read the yamls and use them to mock the outgoing calls during tests. 

Please correct me if I misunderstood the use case.

**Reply by lalitpagaria** [reply]
Yeah kind of similar only thing is how to send api (along with request and response) to keploy instead keploy recording it.

**Reply by slayerjain** [reply]
@lalitpagaria What would you like keploy to do with the request/responses if not record them?

**Reply by lalitpagaria** [reply]
As mentioned like to generate mocks with the identification of noisy fields.
We have many external integrations almost 50-100, it is difficult to test them.

**Reply by slayerjain** [reply]
@lalitpagaria currently keploy only denoises responses against the server where keploy is integrated. The integrations are likely external servers for you. You can record their responses via keploy, and save them as YAML files. To read them you would need a python integration. Are you planning to parse the generated YAML in python yourself?

**Reply by lalitpagaria** [reply]
If it is straightforward then I could contribute to the Python SDK.

**Reply by slayerjain** [reply]
@lalitpagaria I'm sorry for the late reply. To add support for mocking HTTP calls we would need to: 

1.  Intercept HTTP calls from clients. For eg, while working on the js/ts SDK we discovered there are libraries that already do it like https://github.com/mswjs/interceptors 
2. send the payload along with some metadata (I can share the metadata, it's a few fields like naming the integration, or overriding the path to save the file). We use gRPC, so you can simply generate the python client from the proto file and call the keploy server. 
3. Fetch it back (again via gRPC) when you are testing and then use the interceptor to mock it instead of sending the request to the real external service.

I'm more than happy to support you in adding support over the forum, chat or video call :)

**Reply by slayerjain** [reply]
In python maybe we could use this - https://httpretty.readthedocs.io/en/latest/introduction.html

**Comment by lalitpagaria** [comment]
Adding another https://mitmproxy.org/ which have Python API as well.
